---
output: github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# rhinotypeR

<!-- badges: start -->
<!-- badges: end -->

The goal of rhinotypeR package is to perform genotyping on rhinovirus sequences. 

- The package does the following:
  * Assigns genotypes to query sequences
  * Computes for pairwise distance among query sequences
  * Calculates pairwise distance between query and prototype sequences
  * Calculates overall genetic distance of query sequences

## Test-Data 
The project utilizes VP4/2 sequences available in the public domain from GenBank and reference prototype strains from www.picornaviridae.com  

#### Data Format
The input datasets (target, reference and prototype) are fasta files.
Here's an example of a FASTA file:
![fasta file]("https://github.com/omicscodeathon/rhinotyper/blob/main/rhinotypeR/man/figures/example_fasta_file.png")

## Installation

You can install the development version of rhinotypeR from [GitHub](https://github.com) with:

``` r
devtools::install_github("omicscodeathon/rhinotypeR")
```

### Load Library
```{r }
library(rhinotypeR)
```


## Functions
The package encompasses functions to compute genetic distances, perform phylogenetic clustering, and compare sequences against RV prototype strains. 
These functionalities are designed to be user-friendly and adaptable to various research needs.

The list of functions and their roles are summarized in Table 1.

| Function        | Role                                   | Input                    | Output                     
|-----------------|----------------------------------------|--------------------------|-----------------------------
| `getPrototypeSeqs()`| Creates an output directory for the prototype files | fasta file | output directory
| `readFasta`()` | Assigns genotypes to query sequence | fasta file | CSV file
| `SNPeek()` | Assigns genotypes to query sequence | fasta file | CSV file
| `assignTypes()` | Assigns genotypes to query sequence | fasta file | CSV file  
| `allPrototypeDistances()` | Generates pairwise distance between query and prototype sequences | fasta file |
| `pairwiseDistances()` | Assigns genotypes to query sequence | fasta file | CSV file
| `overallMeanDistance()` | Calculates overall genetic distance of query sequences  |  fasta file | 
| `countSNPs`()` | Assigns genotypes to query sequence | fasta file | CSV file
| `PlotFrequency()` | Plot the assigned genotypes | output from assignTypes | Barplot
| `PlotPrototypeDistances()` | Plots prototype distances | distance matrix from prototype distance function | Heatmap
| `PlotTree()` | Plot the assigned genotypes | output from assignTypes | Barplot


### Running the functions using our test data

##### Function 1: getPrototypeSeqs 
- Create an output directory for the prototype files
- Ensure you create an output directory in your current directory before running this function, otherwise it will throw an error saying "output" folder is not found.

Example
```{r }
getPrototypeSeqs(destinationFolder = "./output")
```

Own data
```{r }
#getPrototypeSeqs(destinationFolder = "path to an output folder")
```

##### Function 2: readFasta 
- Read sequences from a FASTA file

Example
```{r }
readFasta(RVAPrototype)
```

Own data
```{r }
#readFasta("path to fastaFile")
```

##### Function 3: SNPeek 
- Visualizing Single Nucleotide Polymorphisms (SNPs) in the fasta file

Example
```{r }
SNPeek(readFasta(RVAPrototype))
```

Own data
```{r }
#SNPeek(readFasta("path to fasta file"))
```

##### Function 4: assignTypes - Assign genotypes to the query sequence
- The model could be either p-distance, JC, Kimura2p, Tamura3p based on your preference.
- The threshhold can also be changed based on your needs.

Example
```{r }
assignTypes(RVBPrototype, readFasta(target_fasta_file), "p-distance", 0.105)
```

Own data
```{r }
#assignTypes("path to reference sequence e.g RVBPrototype", readFasta("path to query sequence"), model, threshold = 0.105)
```

##### Function 4: allPrototypeDistances 
- Calculates pairwise distance between query and prototype sequences
- The model could be either p-distance, JC, Kimura2p, Tamura3p based on your preference.

Example
```{r }
allPrototypeDistances(RVBPrototype, readFasta(target_fasta_file), "p-distance")
```

Own data
```{r }
#allPrototypeDistances("path to reference sequence e.g RVBPrototype", readFasta("path to query sequence"), "Tamura3p")
```

##### Function 5: pairwiseDistances 
- Calculate pairwise distance among input sequences

Example
```{r }
pairwiseDistances(RVBPrototype, "p-distance")
```

Own data
```{r }
#pairwiseDistances("path to prototype file", model = "p-distance")
```

##### Function 6: overallMeanDistance 
- Calculates the overall genetic distance of query sequences

Example
```{r }
overallMeanDistance(readFasta(RVAPrototype), model="p-distance")
```

Own data
```{r }
#overallMeanDistance(readFasta("path to reference sequence e.g RVAPrototype"),  model="p-distance")
```

##### Function 7: countSNPs 
- Count SNPs among input sequences

Example
```{r }
countSNPs(inputSequencesPath = RVBPrototypeAligned)
```

Own data
```{r }
#countSNPs(inputSequencesPath = "path to prototype file") 
```
##### Function 8: Plot frequency
- Create a barplot of genotype frequencies

Example
```{r }
plotFrequency(target_fasta_2, refSeq, "Tamura3p")
```

Own data
```{r }
#plotFrequency("target_data", "ref_data", model = "Tamura3p")
```

##### Function 9: plotPrototypeDistances 
- Visualize genetic distances of the query to prototype sequences
- The output of the generate pairwise distance between query and prototype sequences (function 4) is the input data.

Example
```{r }
plotPrototypeDistances(RVBPrototype, "p-distance")
```

Own data
```{r }
#plotPrototypeDistances("prototypefile", model = "p-distance")
```

##### Function 10: plotTree 
- Plot a simple phylogenetic tree based on distances

Example
```{r }
plotTree(RVBPrototype, "p-distance")
```

Own data
```{r }
#plotTree(prototypefile, model = "p-distance")
```