# Rhinotyper: A package for rhinovirus ghenotyping

## Table of Contents
1. [Background](#Background)
2. [Package](#Package)
3. [Citation](#Installation)
4. [Contributors](#Contributors)

## Background
Rhinoviruses (RV), common respiratory pathogens, are positive-sense, single-stranded RNA viruses characterized by a high antigenic diversity and mutation rate. With their genome approximately 7.2 kb in length, RVs exhibit mutation rates between 10^-3 and 10^-5 mutations per nucleotide per replication event. These viruses are classified into 169 types across three species: RV-A, RV-B, and RV-C. Genotype assignment, a critical aspect of RV research, is based on pairwise genetic distances and phylogenetic clustering with prototype strains, a process currently executed manually and laboriously.   

## Package
Our project aims to develop an R package to automate RV genotype assignment, facilitating genomic scientists in efficiently genotyping RV infections.

Our methodology involves:
1. Parsing and preprocessing of VP4/2 sequence data.
2. Implementation of algorithms to calculate pairwise genetic distances.
3. Integration of methods for constructing Maximum Likelihood phylogenetic 
trees.

More details on the package can be accessed here: [rhinotypeR](https://github.com/omicscodeathon/rhinotyper/tree/main/rhinotypeR)

## Citation

## Contributors

   - Martha Luka

   - Winfred Gatua

   - Wafaa Rashed

   - Ruth Nanjala
