---
output: github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# rhinotypeR

<!-- badges: start -->
<!-- badges: end -->

The goal of rhinotypeR package is to perform genotyping on rhinovirus sequences. 

- The package does the following:
  * Assigns genotypes to query sequences
  * Computes for pairwise distance among query sequences
  * Calculates pairwise distance between query and prototype sequences
  * Calculates overall genetic distance of query sequences

## Test-Data 
The project utilizes VP4/2 sequences available in the public domain from GenBank and reference prototype strains from www.picornaviridae.com  

#### Data Format
The input datasets (target, reference and prototype) are fasta files.
Here's an example of a FASTA file:
![fasta file]("https://github.com/omicscodeathon/rhinotyper/blob/main/rhinotypeR/man/figures/example_fasta_file.png")

## Installation

You can install the development version of rhinotypeR from [GitHub](https://github.com) with:

``` r
devtools::install_github("omicscodeathon/rhinotypeR")
```

## Functions
The package encompasses functions to compute genetic distances, perform phylogenetic clustering, and compare sequences against RV prototype strains. 
These functionalities are designed to be user-friendly and adaptable to various research needs.

The list of functions and their roles are summarized in Table 1.

| Function        | Role                                   | Input                    | Output                     
|-----------------|----------------------------------------|--------------------------|-----------------------------
| `assignTypes()` | Assigns genotypes to query sequence | fasta file | CSV file  
| `PlotFrequency()` | Plot the assigned genotypes | output from assignTypes | Barplot
| `getPrototypeSeqs()`| Creates an output directory for the prototype files | fasta file |                    
| `allPrototypeDistances()` | Generates pairwise distance between query and prototype sequences | fasta file |
| `PlotPrototypeDistances()` | Plots prototype distances | distance matrix from prototype distance function | Heatmap
| `overallMeanDistance()` | Calculates overall genetic distance of query sequences  |  fasta file |          

### Running the functions using our test data

##### Load Library

```{r }
library(rhinotypeR)
```

##### Function 1: Assign genotypes to query sequence
```{r }
assignTypes(pathToRef, pathToQuery, model = "p-distance", threshold = 0.105)
```
##### Function 2: Plot frequency
```{r }
PlotFrequency(genotypeassigned)
```

##### Function 3: Create an output directory for the prototype files
```{r }
#Create an output directory in your current directory before running this function
getPrototypeSeqs(destinationFolder = "./output")
```

##### Function 4: Generate pairwise distance between query and prototype sequences
```{r }
allPrototypeDistances(RVBPrototype, pathToQuery, "p-distance")
```

##### Function 5: Plot Prototype distances
```{r }
#PlotPrototypeDistances(pairwisedist)
```

##### Function 6: Calculate overall genetic distance of query sequences
```{r }
overallMeanDistance(RVAPrototype)
```

### Running the functions on your own dataset

##### Load library

```{r }
library(Rhinotyper)
```

##### Function 1: Assign genotypes to query sequence
The model could be either p-distance, JC, Kimura2p, Tamura3p based on your preference.
The threshhold can also be changed based on your needs.

```{r }
#assignTypes("path to reference sequence e.g RVBPrototype", "path to query sequence", model = "Tamura3p", threshold = 0.105)
```

##### Function 2: Plot frequency
The output of the assign genotype (function 1) is the input.
```{r }
#genotypeassigned <- assignTypes(pathToRef, pathToQuery, model = "p-distance", threshold = 0.105)
#PlotFrequency(genotypeassigned)
```

##### Function 3: Create an output directory for the prototype files
Ensure you create an output directory in your current directory before running this function, otherwise it will throw an error saying "output" folder is not found.

```{r }
getPrototypeSeqs(destinationFolder = "./output")
```

##### Function 4: Generate pairwise distance between query and prototype sequences
The model could be either p-distance, JC, Kimura2p, Tamura3p based on your preference.

```{r }
#allPrototypeDistances("path to reference sequence e.g RVBPrototype", "path to query sequence", "Tamura3p")
```

##### Function 5: Plot Prototype distances
The output of the generate pairwise distance between query and prototype sequences (function 4) is the input data.
```{r }
#PlotPrototypeDistances("output from 'allPrototypeDistances'")
```

##### Function 6: Calculate overall genetic distance of query sequences

```{r }
#overallMeanDistance("path to reference sequence e.g RVAPrototype")
```
