---
output: github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# rhinotypeR

<!-- badges: start -->
<!-- badges: end -->

```

           /**       /**                       /**                                   /*** *** 
          | **      |**/                      | **                                  | **__  **
  /****** | *******  /** /*******   /******  /******   /**   /**  /******   /****** | **  \ **
 /**__  **| **__  **| **| **__  ** /**__  **|_  **_/  | **  | ** /**__  ** /**__  **| *******/
| **  \__/| **  \ **| **| **  \ **| **  \ **  | **    | **  | **| **  \ **| ********| **__  **
| **      | **  | **| **| **  | **| **  | **  | ** /**| **  | **| **  | **| **_____/| **  \ **
| **      | **  | **| **| **  | **|  ******/  |  *****/|  *******/| ********|  *******/| **  | 
|__/      |__/  |__/|__/|__/  |__/ \______/    \___/   \____  **| **____/  \_______/|__/  |__/
                                                       /**  | **| **                          
                                                      |  ******/| **                          
                                                       \______/ |__/                          

```

## Table of Contents
1. [Background](#Background)
2. [Test-Data](#Test-Data)
3. [Workflow](#Workflow)
4. [Package](#Package)
5. [Citation](#Citation)
6. [Contributors](#Contributors)

## Background
Rhinoviruses (RV), common respiratory pathogens, are positive-sense, single-stranded RNA viruses characterized by a high antigenic diversity and mutation rate. With their genome approximately 7.2 kb in length, RVs exhibit mutation rates between 10^-3 and 10^-5 mutations per nucleotide per replication event. These viruses are classified into 169 types across three species: RV-A, RV-B, and RV-C. Genotype assignment, a critical aspect of RV research, is based on pairwise genetic distances and phylogenetic clustering with prototype strains, a process currently executed manually and laboriously. 


## Test-Data
The project utilizes VP4/2 sequences available in the public domain from
GenBank and reference prototype strains from www.picornaviridae.com
The input datasets (target, reference and prototype) are fasta files.
Hereâ€™s an example of a FASTA file: 
![fasta file](https://github.com/omicscodeathon/rhinotyper/blob/main/man/figures/example_fasta_file.png)

## Workflow
![workflow](https://github.com/omicscodeathon/rhinotyper/blob/main/man/figures/workflow.png)


RhinotypeR workflow. The user downloads prototype strains using `getPrototypeSeqs()` function, combines these with their newly generated VP4/2 sequences, aligns and manually curates the alignment. 
The user then reads the curated alignment into R using `readFasta()` function. The readFasta object can then be used to run all the second-level functions, including `assignTypes()` which assigns the sequences into genotypes, filters out the prototype sequences and returns the genotype assignment of the new sequences. This output can be used to visualise the frequency of assigned genotypes.
The distance matrix object, an output of `pairwiseDistance()` function, can be used to create a phylogenetic tree or a heatmap to visualize genetic relatedness of sequences.  


## Package
Our project aims to develop an R package to automate RV genotype assignment, facilitating genomic scientists in efficiently genotyping RV infections.

Our methodology involves:
1. Parsing and preprocessing of VP4/2 sequence data.
2. Implementation of algorithms to calculate pairwise genetic distances.
3. Integration of methods for constructing Maximum Likelihood phylogenetic 
trees.

### Installation

You can install the development version of rhinotypeR from
[GitHub](https://github.com) with:


``` {r, eval=FALSE}
devtools::install_github("omicscodeathon/rhinotypeR")
```

##### Load Library

``` {r }
library("rhinotypeR")
```


## Functions
The package encompasses functions to compute genetic distances, perform phylogenetic clustering, and compare sequences against RV prototype strains. 
These functionalities are designed to be user-friendly and adaptable to various research needs.

- The package (summarized in Table 1) does the following:
  - Assigns genotypes to query sequences
  - Computes for pairwise distance among query sequences
   - Calculates pairwise distance between query and prototype sequences
  - Calculates overall genetic distance of query sequences
 
#### Table 1. A summary of the functions

| Function        | Role                                   | Input                    | Output                     
|-----------------|----------------------------------------|--------------------------|-----------------------------
| `getPrototypeSeqs()`| Downloads RV prototypes into a user-specified local directory | Destination path | RV prototypes are downloaded into the local machine
| `readFasta`()` | Read sequences from a FASTA file | fasta file | A fasta file imported into R
| `SNPeek()` | Visualizing Single Nucleotide Polymorphisms (SNPs) in the fasta file | fasta file | A plot highlighting SNPs per sequence
| `plotAA()` | Visualise amino acid substitutions using a user-specified sequence as the reference | Amino acid fasta file | A plot highlighting amino acid substitutions per sequence
| `assignTypes()` | Assigns genotypes to query sequence | fasta file | CSV file with three columns: sequence header, assigned type, and genetic distance  
| `pairwiseDistances()` | Calculate pairwise distance among input sequences using a user-specified evolutionary model | fasta file | A dense distance matrix
| `overallMeanDistance()` | Calculates the overall mean genetic distance of query sequences using a user-specified evolutionary model  |  fasta file |  A single numeric value
| `countSNPs`()` | Count pairwise SNPs among query sequences | fasta file | A dense matrix
| `PlotFrequency()` | Create a barplot of genotype frequencies | output from assignTypes | Barplot
| `PlotDistances()` | Plots prototype distances | distance matrix from prototype distance function | Heatmap
| `PlotTree()` | Plot a simple phylogenetic tree based on distances | output from pairwise distances | A simple phylogenetic tree


### Running the functions

#### Function 1: getPrototypeSeqs 
- Create an output directory for the prototype files
- Ensure you create an output directory in your current directory before running this function, otherwise it will throw an error saying "output" folder is not found.


Example
```{r }
getPrototypeSeqs(destinationFolder = "./output")
```

Own data
```{r }
# getPrototypeSeqs(destinationFolder = "path to an output folder")
```

##### Function 2: readFasta 
- Read sequences from a FASTA file

Example
```{r }
   # Load the dataset
  test <- system.file("extdata", "test.fasta", package = "rhinotypeR")
  
  # run command
  readFasta(test, desiredLength = 430)
```

Own data
```{r }
# readFasta("path to fastaFile", desiredLength = 430)
```

##### Function 3: SNPeek 
- Visualizing Single Nucleotide Polymorphisms (SNPs) in the fasta file

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "test.fasta", package = "rhinotypeR")
  #read input
  fastaData <- readFasta(fastaFile = test, desiredLength = 480)
  # run command
  SNPeek(fastaData, showLegend = F)
```

Own data
```{r }
  # fastaData <- readFasta(fastaFile = "path to fasta file", desiredLength = 480)
  # SNPeek(fastaData, showLegend = F)
```

##### Function 4: assignTypes - Assign genotypes to the query sequence
- The model could be either p-distance, JC, Kimura2p, Tamura3p based on your preference.
- The threshhold can also be changed based on your needs.

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")
  #read input
  fastaD <- readFasta(test)
  # run command
  head(assignTypes(fastaD, "p-distance"))
```

Own data
```{r }
  # fastaD <- readFasta("path to fasta file")
  # assignTypes(fastaD, "p-distance")
```


##### Function 5: pairwiseDistances 
- Calculate pairwise distance among input sequences

Example
```{r }
# Load the dataset
  test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")
  
  # Example usage
  fastaD <- readFasta(test)
  
  #head(pairwiseDistances(fastaD, "p-distance", pairwiseDeletion = FALSE))
  #head(pairwiseDistances(fastaD, "JC", pairwiseDeletion = FALSE))
  #head(pairwiseDistances(fastaD, "Kimura2p", pairwiseDeletion = FALSE))
  #head(pairwiseDistances(fastaD, "Tamura3p", pairwiseDeletion = FALSE))
```

Own data
```{r }
# fastaD <- readFasta("path to fasta file")
# pairwiseDistances(fastaD, model = "p-distance", pairwiseDeletion = FALSE)
```

##### Function 6: overallMeanDistance 
- Calculates the overall genetic distance of query sequences

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")
  
  # usage
  fastaData <- readFasta(test)
  
  overallMeanDistance(fastaData, model="p-distance")
  overallMeanDistance(fastaData, model="JC")
  overallMeanDistance(fastaData, model="Kimura2p")
  overallMeanDistance(fastaData, model="Tamura3p")
```

Own data
```{r }
# fastaD <- readFasta("path to fasta file")
# overallMeanDistance(fastaD,  model="p-distance")
```

##### Function 7: countSNPs 
- Count SNPs among input sequences

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "test.fasta", package = "rhinotypeR")
  
  # run
  fastaData <- readFasta(test)
  countSNPs(fastaData)
```

Own data
```{r }
# fastaD <- readFasta("path to fasta file")
# countSNPs(fastaD) 
```

##### Function 8: plotFrequency
- Create a barplot of genotype frequencies

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")
  
  # Run 
  queryFastaData <- readFasta(test)
  df <- assignTypes(queryFastaData, "p-distance")
  
  plotFrequency(df)
  
  # Color legend
  # "A" = "blue", "B" = "red", "C" = "green", "Other" = "grey"
```

Own data
```{r }
  #queryFastaData <- readFasta("path to fasta file")
  #df <- assignTypes(queryFastaData, "p-distance")
  #plotFrequency(df)
```

##### Function 9: plotDistances 
- Visualize genetic distances of the query to prototype sequences
- The output of the generate pairwise distance between query and prototype sequences (function 4) is the input data.

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")
  
  # Example usage
  fastaD <- readFasta(test)
  distancesToPrototypes <- pairwiseDistances(fastaD, "p-distance")
  plotDistances(distancesToPrototypes)
```

Own data
```{r }
  #fastaD <- readFasta("path to fasta file")
  #distancesToPrototypes <- pairwiseDistances(fastaD, "p-distance")
  #plotDistances(distancesToPrototypes)
```

##### Function 10: plotTree 
- Plot a simple phylogenetic tree based on distances

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")
  
  # Example usage
  fastaD <- readFasta(test)
  pdistances <- pairwiseDistances(fastaD, "p-distance")
  plotTree(pdistances)
```

Own data
```{r }
 #fastaD <- readFasta("path to fasta file")
 #pdistances <- pairwiseDistances(fastaD, "p-distance")
 #plotTree(pdistances)
```


##### Function 11: plotAA 
- Plot a simple phylogenetic tree based on distances

Example
```{r }
  # Load the dataset
  test <- system.file("extdata", "test.fasta", package = "rhinotypeR")
  
  # usage
  fastaData <- readFasta(fastaFile = test, desiredLength = 170)
  plotAA(fastaData, showLegend = T)
  
  # Legend
   # R = "red", H = "red", K = "red",      # Positively charged amino acid 
   # D = "blue", E = "blue",               # Negatively charged amino acid 
   # S = "green", T = "green", N = "green", Q = "green", # Polar amino acid 
   # A = "yellow", V = "yellow", I = "yellow", L = "yellow", M = "yellow", F = "yellow", 
   # W = "yellow", P = "yellow", G = "yellow", Y = "yellow", C = "yellow"  # Nonpolar amino acid 
```

Own data
```{r }
  #fastaD <- readFasta("path to fasta file", desiredLength = 170)
  #plotAA(fastaData, showLegend = T)
```
