---
title: "Introduction to rhinotypeR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to rhinotypeR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

The `rhinotypeR` package is designed to simplify the genotyping of rhinoviruses using the VP4/2 genomic region. It streamlines the process by enabling a user to download prototype sequences, calculate genetic pairwise distances, and compare the distances to prototype strains for genotype assignment.It also provides visualization options such as frequency plots and simple phylogenetic trees. 

## Installation

You can install `rhinotypeR` from BioConductor with:

```{r eval=FALSE}
# If needed
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Biostrings")
```

You can install `rhinotypeR` from GitHub with:

```{r}
# If needed
install.packages("devtools")
devtools::install_github("yourusername/rhinotypeR")
```


# Usage

## Loading the package

```{r}
library(rhinotypeR)
```


## Example Workflow

1. Download prototype sequences:

```{r}
# getPrototypeSeqs("path_to_destination_folder")
# For example
# getPrototypeSeqs("~/Desktop")
```

2. Read sequences:

```{r}
# sequences <- readFasta("path/to/your/fasta/file")
```

3. Visualize SNPs:

```{r}
# SNPeek(sequences)
```

4. Calculate Pairwise Distances:

```{r}
# distances <- pairwiseDistances(sequences)
```

5. Assign Genotypes::

```{r}
# genotypes <- assignTypes(sequences)
```

6. Plot Results:

```{r}
# plotFrequency(genotypes)
# plotDistances(distances)
```
## Session Info

```{r}
sessionInfo()
```

## Conclusion
The rhinotypeR package simplifies the process of genotyping rhinoviruses and analyzing their genetic data.
