---
title: "Introduction to rhinotypeR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to rhinotypeR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
options(repos = c(CRAN = "https://cran.rstudio.com/"))

```


# Introduction

The `rhinotypeR` package is designed to simplify the genotyping of rhinoviruses using the VP4/2 genomic region. Having worked on rhinoviruses for a few years, I noticed that assigning genotypes after sequencing was particularly laborious, and needed several manual interventions. So I reached out to a few colleagues with the idea of a collaborative project to address this. This package addresses this challenge by streamlining the process by enabling a user to download prototype sequences, calculate genetic pairwise distances, and compare the distances to prototype strains for genotype assignment. It also provides visualization options such as frequency plots and simple phylogenetic trees. 

## Installation

You can install `rhinotypeR` from BioConductor with:

```{r eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Biostrings")
```

You can install `rhinotypeR` from GitHub with:

```{r}
install.packages("devtools")

devtools::install_github("omicscodeathon/rhinotypeR")
```


# Usage

## Loading the package

```{r}
library(rhinotypeR)
```


## Example Workflow

1. Download prototype sequences:

The `getPrototypeSeqs` function downloads the prototype sequences required for genotyping. These should the be combined with the newly generated sequences, aligned using a suitable software, and imported into R.

```{r eval=FALSE}
getPrototypeSeqs("~/Desktop")
```

2. Read sequences:

The `readFasta` function reads FASTA files containing sequence data.

```{r}
sequences <- readFasta("../inst/extdata/input_aln.fasta")
```

3. Visualize SNPs:

The `SNPeek` function visualizes single nucleotide polymorphisms (SNPs) in the sequences, with a select sequence acting as the reference. To specify the reference sequences, move it to the bottom of the alignment before importing into R.

```{r}
SNPeek(sequences)
```

4. Calculate Pairwise Distances:

The `pairwiseDistances` function calculates genetic distances between sequences, using a specified evolutionary model.

```{r}
distances <- pairwiseDistances(sequences, model = "p-distance", gapDeletion = TRUE)
```

5. Assign Genotypes:

The assignTypes function assigns genotypes to the sequences by comparing genetic distances to prototype strains.

```{r}
genotypes <- assignTypes(sequences, model = "p-distance", gapDeletion = TRUE, threshold = 0.105)
```

6. Plot Results:

The `plotFrequency` function visualizes the frequency of assigned genotypes. This function uses the output of `assignTypes` as input.

```{r}
plotFrequency(genotypes)
plotDistances(distances)
```


The `plotDistances` function visualizes pairwise genetic distances in a heatmap. This function uses the output of `pairwiseDistances` as input.

```{r}
plotDistances(distances)
```

The `plotTree` function plots a simple phylogenetic tree. This function uses the output of `pairwiseDistances` as input.

```{r}
plotTree(distances)
```

## Conclusion
The rhinotypeR package simplifies the process of genotyping rhinoviruses and analyzing their genetic data. By automating various steps and providing visualization tools, it enhances the efficiency and accuracy of rhinovirus epidemiological studies.
