{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ace22d5e-ef1b-4675-950f-3a80a85322b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test reading in a multiple sequence alignment\n",
    "#Perform pairwise distance calculation\n",
    "#How to interprete that"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1ae2f5da-925b-47dc-accd-5efe39f1434d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install and load necessary packages\n",
    "#install.packages('ape','seqinr')\n",
    "library(ape)\n",
    "library(seqinr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7eab75d0-d35e-46e2-bbd8-a71d659788a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                       97412_C43 GQ323774.1_RVC10genome\n",
       "GQ323774.1_RVC10genome 0.4739596                       \n",
       "95114_C10              0.4662524              0.2197935"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read in the alignment (mafft output)\n",
    "dat <- read.alignment('~/Downloads/RhinotypeR/rhinotyper_testcode_build/test_realigned.fasta',format = 'fasta')\n",
    "# Calculate the pairwise distances\n",
    "dist.alignment(dat,matrix='identity')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c3e3a5-6902-4401-8835-8577d04635ae",
   "metadata": {},
   "source": [
    "These functions compute a matrix of pairwise distances from aligned sequences using similarity (Fitch matrix, for protein sequences only) or identity matrix (for protein and DNA sequences). The resulting matrix contains the squared root of the pairwise distances. For example, if identity between 2 sequences is 80 the squared root of (1.0 - 0.8) i.e. 0.4472136. Note: seqinr::dist.alignment is the square root version of ape::dist.gene (and not ape::dist.dna). [seqinr](https://search.r-project.org/CRAN/refmans/seqinr/html/dist.alignment.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "08612c85-c654-4ffa-bb47-79144cccefc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                        97816_A8 98103_B48 97412_C43 GQ323774.1_RVC10genome\n",
       "98103_B48              0.6426846                                           \n",
       "97412_C43              0.5815800 0.6118792                                 \n",
       "GQ323774.1_RVC10genome 0.5815800 0.6040935 0.4739596                       \n",
       "95114_C10              0.5836834 0.6274710 0.4662524              0.2197935"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read in the alignment (mafft output)\n",
    "dat <- read.alignment('~/Downloads/RhinotypeR/rhinotyper_testcode_build/test_align.fasta',format = 'fasta')\n",
    "# Calculate the pairwise distances\n",
    "dist.alignment(dat,matrix='identity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cfdc7e23-5e6d-4afd-ad73-d1542df4d5a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                       97412_C43 GQ323774.1_RVC10genome\n",
       "GQ323774.1_RVC10genome       101                       \n",
       "95114_C10                     90                     28"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#\n",
    "library(ape)\n",
    "mat <- as.matrix(dat)#ape takes in a matrix, hence convert the alignment to a matrix\n",
    "ape::dist.gene(mat,method='pairwise')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b92b24b-785d-472c-9ddf-303b58052441",
   "metadata": {},
   "source": [
    "The distance **d** between two individuals is the number of loci for which they differ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d7387255-6c64-4e2e-b4a9-ca40cb6d9a2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>dist.dna {ape}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2 id='dist.dna'>Pairwise Distances from DNA Sequences</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>This function computes a matrix of pairwise distances from DNA\n",
       "sequences using a model of DNA evolution. Eleven substitution models\n",
       "(and the raw distance) are currently available.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>dist.dna(x, model = \"K80\", variance = FALSE,\n",
       "         gamma = FALSE, pairwise.deletion = FALSE,\n",
       "         base.freq = NULL, as.matrix = FALSE)\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr><td><code id=\"dist.dna_:_x\">x</code></td>\n",
       "<td>\n",
       "<p>a matrix or a list containing the DNA sequences; this must be\n",
       "of class <code>\"DNAbin\"</code> (use <code>as.DNAbin</code> is they are\n",
       "stored as character).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"dist.dna_:_model\">model</code></td>\n",
       "<td>\n",
       "<p>a character string specifying the evolutionary model to be\n",
       "used; must be one of <code>\"raw\"</code>, <code>\"N\"</code>, <code>\"TS\"</code>,\n",
       "<code>\"TV\"</code>, <code>\"JC69\"</code>, <code>\"K80\"</code> (the default),\n",
       "<code>\"F81\"</code>, <code>\"K81\"</code>, <code>\"F84\"</code>, <code>\"BH87\"</code>,\n",
       "<code>\"T92\"</code>, <code>\"TN93\"</code>, <code>\"GG95\"</code>, <code>\"logdet\"</code>,\n",
       "<code>\"paralin\"</code>, <code>\"indel\"</code>, or <code>\"indelblock\"</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"dist.dna_:_variance\">variance</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether to compute the variances\n",
       "of the distances; defaults to <code>FALSE</code> so the variances are not\n",
       "computed.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"dist.dna_:_gamma\">gamma</code></td>\n",
       "<td>\n",
       "<p>a value for the gamma parameter possibly used to apply a\n",
       "correction to the distances (by default no correction is applied).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"dist.dna_:_pairwise.deletion\">pairwise.deletion</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether to delete the\n",
       "sites with missing data in a pairwise way. The default is to delete\n",
       "the sites with at least one missing data for all sequences (ignored\n",
       "if <code>model = \"indel\"</code> or <code>\"indelblock\"</code>).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"dist.dna_:_base.freq\">base.freq</code></td>\n",
       "<td>\n",
       "<p>the base frequencies to be used in the computations\n",
       "(if applicable). By default, the base frequencies are computed from\n",
       "the whole set of sequences.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"dist.dna_:_as.matrix\">as.matrix</code></td>\n",
       "<td>\n",
       "<p>a logical indicating whether to return the results as\n",
       "a matrix. The default is to return an object of class\n",
       "dist.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>The molecular evolutionary models available through the option\n",
       "<code>model</code> have been extensively described in the literature. A\n",
       "brief description is given below; more details can be found in the\n",
       "references.\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p><code>raw</code>, <code>N</code>: This is simply the proportion or the number of\n",
       "sites that differ between each pair of sequences. This may be useful\n",
       "to draw &ldquo;saturation plots&rdquo;. The options <code>variance</code> and\n",
       "<code>gamma</code> have no effect, but <code>pairwise.deletion</code> can.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>TS</code>, <code>TV</code>: These are the numbers of transitions and\n",
       "transversions, respectively.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>JC69</code>: This model was developed by Jukes and Cantor (1969). It\n",
       "assumes that all substitutions (i.e. a change of a base by another\n",
       "one) have the same probability. This probability is the same for all\n",
       "sites along the DNA sequence. This last assumption can be relaxed by\n",
       "assuming that the substition rate varies among site following a\n",
       "gamma distribution which parameter must be given by the user. By\n",
       "default, no gamma correction is applied. Another assumption is that\n",
       "the base frequencies are balanced and thus equal to 0.25.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>K80</code>: The distance derived by Kimura (1980), sometimes referred\n",
       "to as &ldquo;Kimura's 2-parameters distance&rdquo;, has the same underlying\n",
       "assumptions than the Jukes&ndash;Cantor distance except that two kinds of\n",
       "substitutions are considered: transitions (A &lt;-&gt; G, C &lt;-&gt; T), and\n",
       "transversions (A &lt;-&gt; C, A &lt;-&gt; T, C &lt;-&gt; G, G &lt;-&gt; T). They are assumed\n",
       "to have different probabilities. A transition is the substitution of\n",
       "a purine (C, T) by another one, or the substitution of a pyrimidine\n",
       "(A, G) by another one. A transversion is the substitution of a\n",
       "purine by a pyrimidine, or vice-versa. Both transition and\n",
       "transversion rates are the same for all sites along the DNA\n",
       "sequence. Jin and Nei (1990) modified the Kimura model to allow for\n",
       "variation among sites following a gamma distribution. Like for the\n",
       "Jukes&ndash;Cantor model, the gamma parameter must be given by the\n",
       "user. By default, no gamma correction is applied.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>F81</code>: Felsenstein (1981) generalized the Jukes&ndash;Cantor model\n",
       "by relaxing the assumption of equal base frequencies. The formulae\n",
       "used in this function were taken from McGuire et al. (1999).\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>K81</code>: Kimura (1981) generalized his model (Kimura 1980) by\n",
       "assuming different rates for two kinds of transversions: A &lt;-&gt; C and\n",
       "G &lt;-&gt; T on one side, and A &lt;-&gt; T and C &lt;-&gt; G on the other. This is\n",
       "what Kimura called his &ldquo;three substitution types model&rdquo; (3ST), and\n",
       "is sometimes referred to as &ldquo;Kimura's 3-parameters distance&rdquo;.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>F84</code>: This model generalizes K80 by relaxing the assumption\n",
       "of equal base frequencies. It was first introduced by Felsenstein in\n",
       "1984 in Phylip, and is fully described by Felsenstein and Churchill\n",
       "(1996). The formulae used in this function were taken from McGuire\n",
       "et al. (1999).\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>BH87</code>: Barry and Hartigan (1987) developed a distance based\n",
       "on the observed proportions of changes among the four bases. This\n",
       "distance is not symmetric.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>T92</code>: Tamura (1992) generalized the Kimura model by relaxing\n",
       "the assumption of equal base frequencies. This is done by taking\n",
       "into account the bias in G+C content in the sequences. The\n",
       "substitution rates are assumed to be the same for all sites along\n",
       "the DNA sequence.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>TN93</code>: Tamura and Nei (1993) developed a model which assumes\n",
       "distinct rates for both kinds of transition (A &lt;-&gt; G versus C &lt;-&gt;\n",
       "T), and transversions. The base frequencies are not assumed to be\n",
       "equal and are estimated from the data. A gamma correction of the\n",
       "inter-site variation in substitution rates is possible.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>GG95</code>: Galtier and Gouy (1995) introduced a model where the\n",
       "G+C content may change through time. Different rates are assumed for\n",
       "transitons and transversions.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>logdet</code>: The Log-Det distance, developed by Lockhart et\n",
       "al. (1994), is related to BH87. However, this distance is\n",
       "symmetric. Formulae from Gu and Li (1996) are used.\n",
       "<code>dist.logdet</code> in <span class=\"pkg\">phangorn</span> uses a different\n",
       "implementation that gives substantially different distances for\n",
       "low-diverging sequences.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>paralin</code>: Lake (1994) developed the paralinear distance which\n",
       "can be viewed as another variant of the Barry&ndash;Hartigan distance.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>indel</code>: this counts the number of sites where there is an\n",
       "insertion/deletion gap in one sequence and not in the other.\n",
       "</p>\n",
       "</li>\n",
       "<li><p><code>indelblock</code>: same than before but contiguous gaps are\n",
       "counted as a single unit. Note that the distance between <code>-A-</code> and\n",
       "<code>A--</code> is 3 because there are three different blocks of gaps, whereas\n",
       "the &ldquo;indel&rdquo; distance will be 2.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>an object of class dist (by default), or a numeric\n",
       "matrix if <code>as.matrix = TRUE</code>. If <code>model = \"BH87\"</code>, a numeric\n",
       "matrix is returned because the Barry&ndash;Hartigan distance is not\n",
       "symmetric.\n",
       "</p>\n",
       "<p>If <code>variance = TRUE</code> an attribute called <code>\"variance\"</code> is\n",
       "given to the returned object.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Note</h3>\n",
       "\n",
       "<p>If the sequences are very different, most evolutionary distances are\n",
       "undefined and a non-finite value (Inf or NaN) is returned. You may do\n",
       "<code>dist.dna(, model = \"raw\")</code> to check whether some values are\n",
       "higher than 0.75.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Emmanuel Paradis</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Barry, D. and Hartigan, J. A. (1987) Asynchronous distance between\n",
       "homologous DNA sequences. <em>Biometrics</em>, <b>43</b>, 261&ndash;276.\n",
       "</p>\n",
       "<p>Felsenstein, J. (1981) Evolutionary trees from DNA sequences: a\n",
       "maximum likelihood approach. <em>Journal of Molecular Evolution</em>,\n",
       "<b>17</b>, 368&ndash;376.\n",
       "</p>\n",
       "<p>Felsenstein, J. and Churchill, G. A. (1996) A Hidden Markov model\n",
       "approach to variation among sites in rate of evolution.\n",
       "<em>Molecular Biology and Evolution</em>, <b>13</b>, 93&ndash;104.\n",
       "</p>\n",
       "<p>Galtier, N. and Gouy, M. (1995) Inferring phylogenies from DNA\n",
       "sequences of unequal base compositions. <em>Proceedings of the\n",
       "National Academy of Sciences USA</em>, <b>92</b>, 11317&ndash;11321.\n",
       "</p>\n",
       "<p>Gu, X. and Li, W.-H. (1996) Bias-corrected paralinear and LogDet\n",
       "distances and tests of molecular clocks and phylogenies under\n",
       "nonstationary nucleotide frequencies. <em>Molecular Biology and\n",
       "Evolution</em>, <b>13</b>, 1375&ndash;1383.\n",
       "</p>\n",
       "<p>Jukes, T. H. and Cantor, C. R. (1969) Evolution of protein\n",
       "molecules. in <em>Mammalian Protein Metabolism</em>, ed. Munro, H. N.,\n",
       "pp. 21&ndash;132, New York: Academic Press.\n",
       "</p>\n",
       "<p>Kimura, M. (1980) A simple method for estimating evolutionary rates of\n",
       "base substitutions through comparative studies of nucleotide\n",
       "sequences. <em>Journal of Molecular Evolution</em>, <b>16</b>, 111&ndash;120.\n",
       "</p>\n",
       "<p>Kimura, M. (1981) Estimation of evolutionary distances between\n",
       "homologous nucleotide sequences. <em>Proceedings of the National\n",
       "Academy of Sciences USA</em>, <b>78</b>, 454&ndash;458.\n",
       "</p>\n",
       "<p>Jin, L. and Nei, M. (1990) Limitations of the evolutionary parsimony\n",
       "method of phylogenetic analysis. <em>Molecular Biology and\n",
       "Evolution</em>, <b>7</b>, 82&ndash;102.\n",
       "</p>\n",
       "<p>Lake, J. A. (1994) Reconstructing evolutionary trees from DNA and\n",
       "protein sequences: paralinear distances. <em>Proceedings of the\n",
       "National Academy of Sciences USA</em>, <b>91</b>, 1455&ndash;1459.\n",
       "</p>\n",
       "<p>Lockhart, P. J., Steel, M. A., Hendy, M. D. and Penny, D. (1994)\n",
       "Recovering evolutionary trees under a more realistic model of sequence\n",
       "evolution. <em>Molecular Biology and Evolution</em>, <b>11</b>,\n",
       "605&ndash;602.\n",
       "</p>\n",
       "<p>McGuire, G., Prentice, M. J. and Wright, F. (1999). Improved error\n",
       "bounds for genetic distances from DNA sequences. <em>Biometrics</em>,\n",
       "<b>55</b>, 1064&ndash;1070.\n",
       "</p>\n",
       "<p>Tamura, K. (1992) Estimation of the number of nucleotide substitutions\n",
       "when there are strong transition-transversion and G + C-content\n",
       "biases. <em>Molecular Biology and Evolution</em>, <b>9</b>, 678&ndash;687.\n",
       "</p>\n",
       "<p>Tamura, K. and Nei, M. (1993) Estimation of the number of nucleotide\n",
       "substitutions in the control region of mitochondrial DNA in humans and\n",
       "chimpanzees. <em>Molecular Biology and Evolution</em>, <b>10</b>, 512&ndash;526.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>read.GenBank</code>, <code>read.dna</code>,\n",
       "<code>write.dna</code>,  <code>DNAbin</code>,\n",
       "<code>dist.gene</code>, <code>cophenetic.phylo</code>,\n",
       "<code>dist</code>\n",
       "</p>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>ape</em> version 5.7-1 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{dist.dna}{Pairwise Distances from DNA Sequences}{dist.dna}\n",
       "\\keyword{manip}{dist.dna}\n",
       "\\keyword{multivariate}{dist.dna}\n",
       "\\keyword{cluster}{dist.dna}\n",
       "%\n",
       "\\begin{Description}\n",
       "This function computes a matrix of pairwise distances from DNA\n",
       "sequences using a model of DNA evolution. Eleven substitution models\n",
       "(and the raw distance) are currently available.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "dist.dna(x, model = \"K80\", variance = FALSE,\n",
       "         gamma = FALSE, pairwise.deletion = FALSE,\n",
       "         base.freq = NULL, as.matrix = FALSE)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{x}] a matrix or a list containing the DNA sequences; this must be\n",
       "of class \\code{\"DNAbin\"} (use \\code{\\LinkA{as.DNAbin}{as.DNAbin}} is they are\n",
       "stored as character).\n",
       "\\item[\\code{model}] a character string specifying the evolutionary model to be\n",
       "used; must be one of \\code{\"raw\"}, \\code{\"N\"}, \\code{\"TS\"},\n",
       "\\code{\"TV\"}, \\code{\"JC69\"}, \\code{\"K80\"} (the default),\n",
       "\\code{\"F81\"}, \\code{\"K81\"}, \\code{\"F84\"}, \\code{\"BH87\"},\n",
       "\\code{\"T92\"}, \\code{\"TN93\"}, \\code{\"GG95\"}, \\code{\"logdet\"},\n",
       "\\code{\"paralin\"}, \\code{\"indel\"}, or \\code{\"indelblock\"}.\n",
       "\\item[\\code{variance}] a logical indicating whether to compute the variances\n",
       "of the distances; defaults to \\code{FALSE} so the variances are not\n",
       "computed.\n",
       "\\item[\\code{gamma}] a value for the gamma parameter possibly used to apply a\n",
       "correction to the distances (by default no correction is applied).\n",
       "\\item[\\code{pairwise.deletion}] a logical indicating whether to delete the\n",
       "sites with missing data in a pairwise way. The default is to delete\n",
       "the sites with at least one missing data for all sequences (ignored\n",
       "if \\code{model = \"indel\"} or \\code{\"indelblock\"}).\n",
       "\\item[\\code{base.freq}] the base frequencies to be used in the computations\n",
       "(if applicable). By default, the base frequencies are computed from\n",
       "the whole set of sequences.\n",
       "\\item[\\code{as.matrix}] a logical indicating whether to return the results as\n",
       "a matrix. The default is to return an object of class\n",
       "\\LinkA{dist}{dist}.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\n",
       "The molecular evolutionary models available through the option\n",
       "\\code{model} have been extensively described in the literature. A\n",
       "brief description is given below; more details can be found in the\n",
       "references.\n",
       "\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \\code{raw}, \\code{N}: This is simply the proportion or the number of\n",
       "sites that differ between each pair of sequences. This may be useful\n",
       "to draw ``saturation plots''. The options \\code{variance} and\n",
       "\\code{gamma} have no effect, but \\code{pairwise.deletion} can.\n",
       "\n",
       "\\item{} \\code{TS}, \\code{TV}: These are the numbers of transitions and\n",
       "transversions, respectively.\n",
       "\n",
       "\\item{} \\code{JC69}: This model was developed by Jukes and Cantor (1969). It\n",
       "assumes that all substitutions (i.e. a change of a base by another\n",
       "one) have the same probability. This probability is the same for all\n",
       "sites along the DNA sequence. This last assumption can be relaxed by\n",
       "assuming that the substition rate varies among site following a\n",
       "gamma distribution which parameter must be given by the user. By\n",
       "default, no gamma correction is applied. Another assumption is that\n",
       "the base frequencies are balanced and thus equal to 0.25.\n",
       "\n",
       "\\item{} \\code{K80}: The distance derived by Kimura (1980), sometimes referred\n",
       "to as ``Kimura's 2-parameters distance'', has the same underlying\n",
       "assumptions than the Jukes--Cantor distance except that two kinds of\n",
       "substitutions are considered: transitions (A <-> G, C <-> T), and\n",
       "transversions (A <-> C, A <-> T, C <-> G, G <-> T). They are assumed\n",
       "to have different probabilities. A transition is the substitution of\n",
       "a purine (C, T) by another one, or the substitution of a pyrimidine\n",
       "(A, G) by another one. A transversion is the substitution of a\n",
       "purine by a pyrimidine, or vice-versa. Both transition and\n",
       "transversion rates are the same for all sites along the DNA\n",
       "sequence. Jin and Nei (1990) modified the Kimura model to allow for\n",
       "variation among sites following a gamma distribution. Like for the\n",
       "Jukes--Cantor model, the gamma parameter must be given by the\n",
       "user. By default, no gamma correction is applied.\n",
       "\n",
       "\\item{} \\code{F81}: Felsenstein (1981) generalized the Jukes--Cantor model\n",
       "by relaxing the assumption of equal base frequencies. The formulae\n",
       "used in this function were taken from McGuire et al. (1999).\n",
       "\n",
       "\\item{} \\code{K81}: Kimura (1981) generalized his model (Kimura 1980) by\n",
       "assuming different rates for two kinds of transversions: A <-> C and\n",
       "G <-> T on one side, and A <-> T and C <-> G on the other. This is\n",
       "what Kimura called his ``three substitution types model'' (3ST), and\n",
       "is sometimes referred to as ``Kimura's 3-parameters distance''.\n",
       "\n",
       "\\item{} \\code{F84}: This model generalizes K80 by relaxing the assumption\n",
       "of equal base frequencies. It was first introduced by Felsenstein in\n",
       "1984 in Phylip, and is fully described by Felsenstein and Churchill\n",
       "(1996). The formulae used in this function were taken from McGuire\n",
       "et al. (1999).\n",
       "\n",
       "\\item{} \\code{BH87}: Barry and Hartigan (1987) developed a distance based\n",
       "on the observed proportions of changes among the four bases. This\n",
       "distance is not symmetric.\n",
       "\n",
       "\\item{} \\code{T92}: Tamura (1992) generalized the Kimura model by relaxing\n",
       "the assumption of equal base frequencies. This is done by taking\n",
       "into account the bias in G+C content in the sequences. The\n",
       "substitution rates are assumed to be the same for all sites along\n",
       "the DNA sequence.\n",
       "\n",
       "\\item{} \\code{TN93}: Tamura and Nei (1993) developed a model which assumes\n",
       "distinct rates for both kinds of transition (A <-> G versus C <->\n",
       "T), and transversions. The base frequencies are not assumed to be\n",
       "equal and are estimated from the data. A gamma correction of the\n",
       "inter-site variation in substitution rates is possible.\n",
       "\n",
       "\\item{} \\code{GG95}: Galtier and Gouy (1995) introduced a model where the\n",
       "G+C content may change through time. Different rates are assumed for\n",
       "transitons and transversions.\n",
       "\n",
       "\\item{} \\code{logdet}: The Log-Det distance, developed by Lockhart et\n",
       "al. (1994), is related to BH87. However, this distance is\n",
       "symmetric. Formulae from Gu and Li (1996) are used.\n",
       "\\code{dist.logdet} in \\pkg{phangorn} uses a different\n",
       "implementation that gives substantially different distances for\n",
       "low-diverging sequences.\n",
       "\n",
       "\\item{} \\code{paralin}: Lake (1994) developed the paralinear distance which\n",
       "can be viewed as another variant of the Barry--Hartigan distance.\n",
       "\n",
       "\\item{} \\code{indel}: this counts the number of sites where there is an\n",
       "insertion/deletion gap in one sequence and not in the other.\n",
       "\n",
       "\\item{} \\code{indelblock}: same than before but contiguous gaps are\n",
       "counted as a single unit. Note that the distance between \\code{-A-} and\n",
       "\\code{A--} is 3 because there are three different blocks of gaps, whereas\n",
       "the ``indel'' distance will be 2.\n",
       "\n",
       "\\end{itemize}\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "an object of class \\LinkA{dist}{dist} (by default), or a numeric\n",
       "matrix if \\code{as.matrix = TRUE}. If \\code{model = \"BH87\"}, a numeric\n",
       "matrix is returned because the Barry--Hartigan distance is not\n",
       "symmetric.\n",
       "\n",
       "If \\code{variance = TRUE} an attribute called \\code{\"variance\"} is\n",
       "given to the returned object.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Note}\n",
       "If the sequences are very different, most evolutionary distances are\n",
       "undefined and a non-finite value (Inf or NaN) is returned. You may do\n",
       "\\code{dist.dna(, model = \"raw\")} to check whether some values are\n",
       "higher than 0.75.\n",
       "\\end{Note}\n",
       "%\n",
       "\\begin{Author}\n",
       "Emmanuel Paradis\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\n",
       "Barry, D. and Hartigan, J. A. (1987) Asynchronous distance between\n",
       "homologous DNA sequences. \\emph{Biometrics}, \\bold{43}, 261--276.\n",
       "\n",
       "Felsenstein, J. (1981) Evolutionary trees from DNA sequences: a\n",
       "maximum likelihood approach. \\emph{Journal of Molecular Evolution},\n",
       "\\bold{17}, 368--376.\n",
       "\n",
       "Felsenstein, J. and Churchill, G. A. (1996) A Hidden Markov model\n",
       "approach to variation among sites in rate of evolution.\n",
       "\\emph{Molecular Biology and Evolution}, \\bold{13}, 93--104.\n",
       "\n",
       "Galtier, N. and Gouy, M. (1995) Inferring phylogenies from DNA\n",
       "sequences of unequal base compositions. \\emph{Proceedings of the\n",
       "National Academy of Sciences USA}, \\bold{92}, 11317--11321.\n",
       "\n",
       "Gu, X. and Li, W.-H. (1996) Bias-corrected paralinear and LogDet\n",
       "distances and tests of molecular clocks and phylogenies under\n",
       "nonstationary nucleotide frequencies. \\emph{Molecular Biology and\n",
       "Evolution}, \\bold{13}, 1375--1383.\n",
       "\n",
       "Jukes, T. H. and Cantor, C. R. (1969) Evolution of protein\n",
       "molecules. in \\emph{Mammalian Protein Metabolism}, ed. Munro, H. N.,\n",
       "pp. 21--132, New York: Academic Press.\n",
       "\n",
       "Kimura, M. (1980) A simple method for estimating evolutionary rates of\n",
       "base substitutions through comparative studies of nucleotide\n",
       "sequences. \\emph{Journal of Molecular Evolution}, \\bold{16}, 111--120.\n",
       "\n",
       "Kimura, M. (1981) Estimation of evolutionary distances between\n",
       "homologous nucleotide sequences. \\emph{Proceedings of the National\n",
       "Academy of Sciences USA}, \\bold{78}, 454--458.\n",
       "\n",
       "Jin, L. and Nei, M. (1990) Limitations of the evolutionary parsimony\n",
       "method of phylogenetic analysis. \\emph{Molecular Biology and\n",
       "Evolution}, \\bold{7}, 82--102.\n",
       "\n",
       "Lake, J. A. (1994) Reconstructing evolutionary trees from DNA and\n",
       "protein sequences: paralinear distances. \\emph{Proceedings of the\n",
       "National Academy of Sciences USA}, \\bold{91}, 1455--1459.\n",
       "\n",
       "Lockhart, P. J., Steel, M. A., Hendy, M. D. and Penny, D. (1994)\n",
       "Recovering evolutionary trees under a more realistic model of sequence\n",
       "evolution. \\emph{Molecular Biology and Evolution}, \\bold{11},\n",
       "605--602.\n",
       "\n",
       "McGuire, G., Prentice, M. J. and Wright, F. (1999). Improved error\n",
       "bounds for genetic distances from DNA sequences. \\emph{Biometrics},\n",
       "\\bold{55}, 1064--1070.\n",
       "\n",
       "Tamura, K. (1992) Estimation of the number of nucleotide substitutions\n",
       "when there are strong transition-transversion and G + C-content\n",
       "biases. \\emph{Molecular Biology and Evolution}, \\bold{9}, 678--687.\n",
       "\n",
       "Tamura, K. and Nei, M. (1993) Estimation of the number of nucleotide\n",
       "substitutions in the control region of mitochondrial DNA in humans and\n",
       "chimpanzees. \\emph{Molecular Biology and Evolution}, \\bold{10}, 512--526.\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\n",
       "\\code{\\LinkA{read.GenBank}{read.GenBank}}, \\code{\\LinkA{read.dna}{read.dna}},\n",
       "\\code{\\LinkA{write.dna}{write.dna}},  \\code{\\LinkA{DNAbin}{DNAbin}},\n",
       "\\code{\\LinkA{dist.gene}{dist.gene}}, \\code{\\LinkA{cophenetic.phylo}{cophenetic.phylo}},\n",
       "\\code{\\LinkA{dist}{dist}}\n",
       "\\end{SeeAlso}"
      ],
      "text/plain": [
       "dist.dna                  package:ape                  R Documentation\n",
       "\n",
       "_\bP_\ba_\bi_\br_\bw_\bi_\bs_\be _\bD_\bi_\bs_\bt_\ba_\bn_\bc_\be_\bs _\bf_\br_\bo_\bm _\bD_\bN_\bA _\bS_\be_\bq_\bu_\be_\bn_\bc_\be_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     This function computes a matrix of pairwise distances from DNA\n",
       "     sequences using a model of DNA evolution. Eleven substitution\n",
       "     models (and the raw distance) are currently available.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     dist.dna(x, model = \"K80\", variance = FALSE,\n",
       "              gamma = FALSE, pairwise.deletion = FALSE,\n",
       "              base.freq = NULL, as.matrix = FALSE)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "       x: a matrix or a list containing the DNA sequences; this must be\n",
       "          of class ‘\"DNAbin\"’ (use ‘as.DNAbin’ is they are stored as\n",
       "          character).\n",
       "\n",
       "   model: a character string specifying the evolutionary model to be\n",
       "          used; must be one of ‘\"raw\"’, ‘\"N\"’, ‘\"TS\"’, ‘\"TV\"’,\n",
       "          ‘\"JC69\"’, ‘\"K80\"’ (the default), ‘\"F81\"’, ‘\"K81\"’, ‘\"F84\"’,\n",
       "          ‘\"BH87\"’, ‘\"T92\"’, ‘\"TN93\"’, ‘\"GG95\"’, ‘\"logdet\"’,\n",
       "          ‘\"paralin\"’, ‘\"indel\"’, or ‘\"indelblock\"’.\n",
       "\n",
       "variance: a logical indicating whether to compute the variances of the\n",
       "          distances; defaults to ‘FALSE’ so the variances are not\n",
       "          computed.\n",
       "\n",
       "   gamma: a value for the gamma parameter possibly used to apply a\n",
       "          correction to the distances (by default no correction is\n",
       "          applied).\n",
       "\n",
       "pairwise.deletion: a logical indicating whether to delete the sites\n",
       "          with missing data in a pairwise way. The default is to delete\n",
       "          the sites with at least one missing data for all sequences\n",
       "          (ignored if ‘model = \"indel\"’ or ‘\"indelblock\"’).\n",
       "\n",
       "base.freq: the base frequencies to be used in the computations (if\n",
       "          applicable). By default, the base frequencies are computed\n",
       "          from the whole set of sequences.\n",
       "\n",
       "as.matrix: a logical indicating whether to return the results as a\n",
       "          matrix. The default is to return an object of class dist.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     The molecular evolutionary models available through the option\n",
       "     ‘model’ have been extensively described in the literature. A brief\n",
       "     description is given below; more details can be found in the\n",
       "     references.\n",
       "\n",
       "        • ‘raw’, ‘N’: This is simply the proportion or the number of\n",
       "          sites that differ between each pair of sequences. This may be\n",
       "          useful to draw ``saturation plots''. The options ‘variance’\n",
       "          and ‘gamma’ have no effect, but ‘pairwise.deletion’ can.\n",
       "\n",
       "        • ‘TS’, ‘TV’: These are the numbers of transitions and\n",
       "          transversions, respectively.\n",
       "\n",
       "        • ‘JC69’: This model was developed by Jukes and Cantor (1969).\n",
       "          It assumes that all substitutions (i.e. a change of a base by\n",
       "          another one) have the same probability. This probability is\n",
       "          the same for all sites along the DNA sequence. This last\n",
       "          assumption can be relaxed by assuming that the substition\n",
       "          rate varies among site following a gamma distribution which\n",
       "          parameter must be given by the user. By default, no gamma\n",
       "          correction is applied. Another assumption is that the base\n",
       "          frequencies are balanced and thus equal to 0.25.\n",
       "\n",
       "        • ‘K80’: The distance derived by Kimura (1980), sometimes\n",
       "          referred to as ``Kimura's 2-parameters distance'', has the\n",
       "          same underlying assumptions than the Jukes-Cantor distance\n",
       "          except that two kinds of substitutions are considered:\n",
       "          transitions (A <-> G, C <-> T), and transversions (A <-> C, A\n",
       "          <-> T, C <-> G, G <-> T). They are assumed to have different\n",
       "          probabilities. A transition is the substitution of a purine\n",
       "          (C, T) by another one, or the substitution of a pyrimidine\n",
       "          (A, G) by another one. A transversion is the substitution of\n",
       "          a purine by a pyrimidine, or vice-versa. Both transition and\n",
       "          transversion rates are the same for all sites along the DNA\n",
       "          sequence. Jin and Nei (1990) modified the Kimura model to\n",
       "          allow for variation among sites following a gamma\n",
       "          distribution. Like for the Jukes-Cantor model, the gamma\n",
       "          parameter must be given by the user. By default, no gamma\n",
       "          correction is applied.\n",
       "\n",
       "        • ‘F81’: Felsenstein (1981) generalized the Jukes-Cantor model\n",
       "          by relaxing the assumption of equal base frequencies. The\n",
       "          formulae used in this function were taken from McGuire et al.\n",
       "          (1999).\n",
       "\n",
       "        • ‘K81’: Kimura (1981) generalized his model (Kimura 1980) by\n",
       "          assuming different rates for two kinds of transversions: A\n",
       "          <-> C and G <-> T on one side, and A <-> T and C <-> G on the\n",
       "          other. This is what Kimura called his ``three substitution\n",
       "          types model'' (3ST), and is sometimes referred to as\n",
       "          ``Kimura's 3-parameters distance''.\n",
       "\n",
       "        • ‘F84’: This model generalizes K80 by relaxing the assumption\n",
       "          of equal base frequencies. It was first introduced by\n",
       "          Felsenstein in 1984 in Phylip, and is fully described by\n",
       "          Felsenstein and Churchill (1996). The formulae used in this\n",
       "          function were taken from McGuire et al. (1999).\n",
       "\n",
       "        • ‘BH87’: Barry and Hartigan (1987) developed a distance based\n",
       "          on the observed proportions of changes among the four bases.\n",
       "          This distance is not symmetric.\n",
       "\n",
       "        • ‘T92’: Tamura (1992) generalized the Kimura model by relaxing\n",
       "          the assumption of equal base frequencies. This is done by\n",
       "          taking into account the bias in G+C content in the sequences.\n",
       "          The substitution rates are assumed to be the same for all\n",
       "          sites along the DNA sequence.\n",
       "\n",
       "        • ‘TN93’: Tamura and Nei (1993) developed a model which assumes\n",
       "          distinct rates for both kinds of transition (A <-> G versus C\n",
       "          <-> T), and transversions. The base frequencies are not\n",
       "          assumed to be equal and are estimated from the data. A gamma\n",
       "          correction of the inter-site variation in substitution rates\n",
       "          is possible.\n",
       "\n",
       "        • ‘GG95’: Galtier and Gouy (1995) introduced a model where the\n",
       "          G+C content may change through time. Different rates are\n",
       "          assumed for transitons and transversions.\n",
       "\n",
       "        • ‘logdet’: The Log-Det distance, developed by Lockhart et al.\n",
       "          (1994), is related to BH87. However, this distance is\n",
       "          symmetric. Formulae from Gu and Li (1996) are used.\n",
       "          ‘dist.logdet’ in ‘phangorn’ uses a different implementation\n",
       "          that gives substantially different distances for\n",
       "          low-diverging sequences.\n",
       "\n",
       "        • ‘paralin’: Lake (1994) developed the paralinear distance\n",
       "          which can be viewed as another variant of the Barry-Hartigan\n",
       "          distance.\n",
       "\n",
       "        • ‘indel’: this counts the number of sites where there is an\n",
       "          insertion/deletion gap in one sequence and not in the other.\n",
       "\n",
       "        • ‘indelblock’: same than before but contiguous gaps are\n",
       "          counted as a single unit. Note that the distance between\n",
       "          ‘-A-’ and ‘A--’ is 3 because there are three different blocks\n",
       "          of gaps, whereas the ``indel'' distance will be 2.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     an object of class dist (by default), or a numeric matrix if\n",
       "     ‘as.matrix = TRUE’. If ‘model = \"BH87\"’, a numeric matrix is\n",
       "     returned because the Barry-Hartigan distance is not symmetric.\n",
       "\n",
       "     If ‘variance = TRUE’ an attribute called ‘\"variance\"’ is given to\n",
       "     the returned object.\n",
       "\n",
       "_\bN_\bo_\bt_\be:\n",
       "\n",
       "     If the sequences are very different, most evolutionary distances\n",
       "     are undefined and a non-finite value (Inf or NaN) is returned. You\n",
       "     may do ‘dist.dna(, model = \"raw\")’ to check whether some values\n",
       "     are higher than 0.75.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Emmanuel Paradis\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Barry, D. and Hartigan, J. A. (1987) Asynchronous distance between\n",
       "     homologous DNA sequences. _Biometrics_, *43*, 261-276.\n",
       "\n",
       "     Felsenstein, J. (1981) Evolutionary trees from DNA sequences: a\n",
       "     maximum likelihood approach. _Journal of Molecular Evolution_,\n",
       "     *17*, 368-376.\n",
       "\n",
       "     Felsenstein, J. and Churchill, G. A. (1996) A Hidden Markov model\n",
       "     approach to variation among sites in rate of evolution.\n",
       "     _Molecular Biology and Evolution_, *13*, 93-104.\n",
       "\n",
       "     Galtier, N. and Gouy, M. (1995) Inferring phylogenies from DNA\n",
       "     sequences of unequal base compositions. _Proceedings of the\n",
       "     National Academy of Sciences USA_, *92*, 11317-11321.\n",
       "\n",
       "     Gu, X. and Li, W.-H. (1996) Bias-corrected paralinear and LogDet\n",
       "     distances and tests of molecular clocks and phylogenies under\n",
       "     nonstationary nucleotide frequencies. _Molecular Biology and\n",
       "     Evolution_, *13*, 1375-1383.\n",
       "\n",
       "     Jukes, T. H. and Cantor, C. R. (1969) Evolution of protein\n",
       "     molecules. in _Mammalian Protein Metabolism_, ed. Munro, H. N.,\n",
       "     pp. 21-132, New York: Academic Press.\n",
       "\n",
       "     Kimura, M. (1980) A simple method for estimating evolutionary\n",
       "     rates of base substitutions through comparative studies of\n",
       "     nucleotide sequences. _Journal of Molecular Evolution_, *16*,\n",
       "     111-120.\n",
       "\n",
       "     Kimura, M. (1981) Estimation of evolutionary distances between\n",
       "     homologous nucleotide sequences. _Proceedings of the National\n",
       "     Academy of Sciences USA_, *78*, 454-458.\n",
       "\n",
       "     Jin, L. and Nei, M. (1990) Limitations of the evolutionary\n",
       "     parsimony method of phylogenetic analysis. _Molecular Biology and\n",
       "     Evolution_, *7*, 82-102.\n",
       "\n",
       "     Lake, J. A. (1994) Reconstructing evolutionary trees from DNA and\n",
       "     protein sequences: paralinear distances. _Proceedings of the\n",
       "     National Academy of Sciences USA_, *91*, 1455-1459.\n",
       "\n",
       "     Lockhart, P. J., Steel, M. A., Hendy, M. D. and Penny, D. (1994)\n",
       "     Recovering evolutionary trees under a more realistic model of\n",
       "     sequence evolution. _Molecular Biology and Evolution_, *11*,\n",
       "     605-602.\n",
       "\n",
       "     McGuire, G., Prentice, M. J. and Wright, F. (1999). Improved error\n",
       "     bounds for genetic distances from DNA sequences. _Biometrics_,\n",
       "     *55*, 1064-1070.\n",
       "\n",
       "     Tamura, K. (1992) Estimation of the number of nucleotide\n",
       "     substitutions when there are strong transition-transversion and G\n",
       "     + C-content biases. _Molecular Biology and Evolution_, *9*,\n",
       "     678-687.\n",
       "\n",
       "     Tamura, K. and Nei, M. (1993) Estimation of the number of\n",
       "     nucleotide substitutions in the control region of mitochondrial\n",
       "     DNA in humans and chimpanzees. _Molecular Biology and Evolution_,\n",
       "     *10*, 512-526.\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘read.GenBank’, ‘read.dna’, ‘write.dna’, ‘DNAbin’, ‘dist.gene’,\n",
       "     ‘cophenetic.phylo’, ‘dist’\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?ape::dist.dna #Get help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d28ccbcd-875f-460f-9ce7-b1276d4ef4eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# R functions\n",
    "\n",
    "# Read in the alignment (mafft output)\n",
    "dat <- read.alignment('~/Downloads/RhinotypeR/rhinotyper_testcode_build/test_align.fasta',format = 'fasta')\n",
    "# Calculate the pairwise distances\n",
    "dist.alignment(dat,matrix='identity')\n",
    "\n",
    "#\n",
    "library(ape)\n",
    "mat <- as.matrix(dat)#ape takes in a matrix, hence convert the alignment to a matrix\n",
    "ape::dist.gene(mat,method='pairwise')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0732d2c-838d-4973-84e9-4d1d9d9b7bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate pairwise distances using seqinr\n",
    "pairdist_seqinr <- function(data,matrix){\n",
    "    library(seqinr)\n",
    "    pd=dist.alignment(data,matrix)\n",
    "    return(pd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "29f3a64f-eacc-4edb-b644-f387e1ddbf10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                       97412_C43 GQ323774.1_RVC10genome\n",
       "GQ323774.1_RVC10genome 0.4739596                       \n",
       "95114_C10              0.4662524              0.2197935"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairdist_seqinr(data=dat,matrix=\"identity\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1c4a6141-0f9c-4de5-a7e9-cd72016cc6aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate pairwise distances using ape\n",
    "## Convert the alignment to a matrix before parsing that to ape\n",
    "pairdist_ape <- function(data,method){\n",
    "    library(ape)\n",
    "    mat= as.matrix(data)\n",
    "    pdape=ape::dist.gene(mat, method)\n",
    "    return(pdape)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b14adb14-24da-4109-93e4-f9b5257719b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                       97412_C43 GQ323774.1_RVC10genome\n",
       "GQ323774.1_RVC10genome       101                       \n",
       "95114_C10                     90                     28"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pairdist_ape(data=dat, method='pairwise')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
