\nonstopmode{}
\documentclass[a4paper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\makeatletter\@ifl@t@r\fmtversion{2018/04/01}{}{\usepackage[utf8]{inputenc}}\makeatother
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package `rhinotypeR'}}
\par\bigskip{\large \today}
\end{center}
\ifthenelse{\boolean{Rd@use@hyper}}{\hypersetup{pdftitle = {rhinotypeR: Rhinovirus genotyping}}}{}
\ifthenelse{\boolean{Rd@use@hyper}}{\hypersetup{pdfauthor = {Martha Luka; Ruth Nanjala; Winfred Gatua; Wafaa M. Rashed; Olaitan Awe}}}{}
\begin{description}
\raggedright{}
\item[Type]\AsIs{Package}
\item[Title]\AsIs{Rhinovirus genotyping}
\item[Version]\AsIs{0.99.0}
\item[Description]\AsIs{``rhinotypeR'' is designed to automate the comparison of sequence data against prototype strains, 
streamlining the genotype assignment process. By implementing predefined pairwise distance thresholds, 
this package makes genotype assignment accessible to researchers and public health professionals. 
This tool enhances our epidemiological toolkit by enabling more efficient surveillance and analysis of 
rhinoviruses (RVs) and other viral pathogens with complex genomic landscapes. 
Additionally, ``rhinotypeR'' supports comprehensive visualization and analysis of 
single nucleotide polymorphisms (SNPs) and amino acid substitutions, 
facilitating in-depth genetic and evolutionary studies.}
\item[License]\AsIs{MIT + file LICENSE}
\item[Encoding]\AsIs{UTF-8}
\item[Imports]\AsIs{Biostrings}
\item[biocViews]\AsIs{Sequencing, Genetics, Phylogenetics}
\item[Depends]\AsIs{R (>= 4.4.0)}
\item[Suggests]\AsIs{knitr, rmarkdown}
\item[VignetteBuilder]\AsIs{knitr}
\item[Vignettes]\AsIs{`Introduction to rhinotypeR` = ``vignettes/rhinotypeR.Rmd''}
\item[URL]\AsIs{}\url{https://github.com/omicscodeathon/rhinotypeR}\AsIs{}
\item[BugReports]\AsIs{}\url{https://github.com/omicscodeathon/rhinotypeR/issues}\AsIs{}
\item[NeedsCompilation]\AsIs{no}
\item[Author]\AsIs{Martha Luka [aut],
Ruth Nanjala [aut, cre],
Winfred Gatua [aut],
Wafaa M. Rashed [aut],
Olaitan Awe [aut]}
\item[Maintainer]\AsIs{Ruth Nanjala }\email{ruthnanjala97@gmail.com}\AsIs{}
\end{description}
\Rdcontents{Contents}
\HeaderA{assignTypes}{Assigns genotypes to input sequences}{assignTypes}
\keyword{genotype}{assignTypes}
\keyword{sequence analysis}{assignTypes}
%
\begin{Description}
Rapidly assigns genotypes to input sequences. 
The input fasta file should include the prototype strains, 
which can be downloaded using getPrototypeSeqs().
\end{Description}
%
\begin{Usage}
\begin{verbatim}
assignTypes(fastaData, model = "p-distance", gapDeletion = TRUE, threshold = 0.105)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fastaData}] 
The fasta data to be processed.

\item[\code{model}] 
The evolutionary model to be used. Default is set to "p-distance".

\item[\code{gapDeletion}] 
Logical. If TRUE, gaps are deleted. Default is TRUE.

\item[\code{threshold}] 
The distance threshold for genotype assignment. Default is 0.105.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function compares input sequences against prototype strains using a 
specified evolutionary model and assigns genotypes based on predefined distance thresholds.
\end{Details}
%
\begin{Value}
A list with the assigned genotypes and their distances to the reference sequences.
\end{Value}
%
\begin{Note}
Ensure the input fasta file includes the necessary prototype strains.
\end{Note}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{getPrototypeSeqs}{getPrototypeSeqs}}, \code{\LinkA{readFasta}{readFasta}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")

# Run command
fastaD <- readFasta(test)
assignTypes(fastaD, model = "p-distance", gapDeletion = TRUE, threshold = 0.105)
\end{ExampleCode}
\end{Examples}
\HeaderA{countSNPs}{Counts single nucleotide polymorphisms}{countSNPs}
\keyword{genotype}{countSNPs}
\keyword{sequence analysis}{countSNPs}
%
\begin{Description}
Counts single nucleotide polymorphisms across input sequences and 
produces an output matrix.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
countSNPs(fastaData, gapDeletion = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fastaData}] 
The fasta data to be processed.

\item[\code{gapDeletion}] 
Whether or not to delete positions with gaps. Default is set to TRUE.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function counts the number of single nucleotide polymorphisms 
(SNPs) across the provided sequences.
\end{Details}
%
\begin{Value}
A matrix with the SNP counts for each sequence.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{readFasta}{readFasta}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "test.fasta", package = "rhinotypeR")

# Run the function
fastaData <- readFasta(test)
countSNPs(fastaData)
\end{ExampleCode}
\end{Examples}
\HeaderA{getPrototypeSeqs}{Download rhinovirus prototype strains}{getPrototypeSeqs}
\keyword{genotype}{getPrototypeSeqs}
\keyword{sequence analysis}{getPrototypeSeqs}
%
\begin{Description}
Download rhinovirus prototype strains into a local directory. 
These sequences should be combined with and aligned alongside 
newly generated sequences before being imported into R for genotype assignment.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getPrototypeSeqs(destinationFolder)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{destinationFolder}] 
Provide a path that will act as an output folder for the prototype sequences. 
The default is set to "output", but you will need to create it 
prior to running the command in your current directory.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function downloads rhinovirus prototype strains and saves them in the specified folder.
\end{Details}
%
\begin{Value}
A character vector of the downloaded file paths.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{assignTypes}{assignTypes}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Run the function
# getPrototypeSeqs(destinationFolder = "~/Desktop")
\end{ExampleCode}
\end{Examples}
\HeaderA{overallMeanDistance}{Estimates the overall mean distance}{overallMeanDistance}
\keyword{genotype}{overallMeanDistance}
\keyword{sequence analysis}{overallMeanDistance}
%
\begin{Description}
Estimates the overall mean distance of input sequences.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
overallMeanDistance(fastaData, model="p-distance", gapDeletion=TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fastaData}] 
The fasta file used here is the output from the function \code{readFasta}.

\item[\code{model}] 
The evolutionary model used to calculate distances. Default is set to "p-distance".

\item[\code{gapDeletion}] 
Whether or not to delete positions with gaps. Default is set to TRUE.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function estimates the overall mean genetic distance of 
input sequences using the specified evolutionary model.
\end{Details}
%
\begin{Value}
A numeric value representing the overall mean distance.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{pairwiseDistances}{pairwiseDistances}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")

# Usage
fastaData <- readFasta(test)
overallMeanDistance(fastaData, model="p-distance")
\end{ExampleCode}
\end{Examples}
\HeaderA{pairwiseDistances}{Estimates pairwise distances}{pairwiseDistances}
\keyword{genotype}{pairwiseDistances}
\keyword{sequence analysis}{pairwiseDistances}
%
\begin{Description}
Estimates pairwise distances across input sequences using a specified evolutionary model.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
pairwiseDistances(fastaData, model = "p-distance", gapDeletion = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fastaData}] 
The fasta data to be processed.

\item[\code{model}] 
The evolutionary model used to calculate distances. Default is set to "p-distance".

\item[\code{gapDeletion}] 
Whether or not to delete positions with gaps. Default is set to TRUE.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function calculates the pairwise genetic distances between sequences using the specified evolutionary model.
\end{Details}
%
\begin{Value}
A matrix of pairwise distances.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{overallMeanDistance}{overallMeanDistance}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")

# Example usage
fastaD <- readFasta(test)
pairwiseDistances(fastaD, "p-distance")
\end{ExampleCode}
\end{Examples}
\HeaderA{plotAA}{Visualize amino acid substitutions}{plotAA}
\keyword{genotype}{plotAA}
\keyword{sequence analysis}{plotAA}
%
\begin{Description}
Plots amino acid substitutions with a specified sequence as the reference. 
The input is an amino acid fasta file (translated DNA sequences). 
To specify the reference sequence, move it to the bottom of the alignment. 
Changes are colored by the class of amino acid: 
Red = Positively charged, Blue = Negatively charged, Green = Polar, Yellow = Non-polar.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotAA(AAfastaFile, showLegend = FALSE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{AAfastaFile}] 
The file path to the input amino acid sequences in fasta format.

\item[\code{showLegend}] 
Logical indicating whether to show the legend. Default is FALSE.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function visualizes amino acid substitutions in a given set of sequences with color-coded classes.
\end{Details}
%
\begin{Value}
A plot object showing the amino acid substitutions.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{SNPeek}{SNPeek}},
\code{\LinkA{plotAA}{plotAA}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "test.translated.fasta", package = "rhinotypeR")

# Usage
plotAA(test)
\end{ExampleCode}
\end{Examples}
\HeaderA{plotDistances}{Visualizes pairwise genetic distances}{plotDistances}
\keyword{genotype}{plotDistances}
\keyword{sequence analysis}{plotDistances}
%
\begin{Description}
Visualizes pairwise genetic distances in a heatmap. 
This function uses the output of pairwiseDistances() as input.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotDistances(distancesMatrix)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{distancesMatrix}] 
A matrix of pairwise genetic distances from the function \code{pairwiseDistances}.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function creates a heatmap to visualize the pairwise genetic distances between sequences.
\end{Details}
%
\begin{Value}
A heatmap plot object.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{pairwiseDistances}{pairwiseDistances}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")

# Example usage
fastaD <- readFasta(test)
distancesMatrix <- pairwiseDistances(fastaD, "p-distance")
plotDistances(distancesMatrix)
\end{ExampleCode}
\end{Examples}
\HeaderA{plotFrequency}{Plots the frequency of assigned genotypes}{plotFrequency}
\keyword{genotype}{plotFrequency}
\keyword{visualization}{plotFrequency}
%
\begin{Description}
Plots the frequency of assigned genotypes. 
This function uses the output of \code{assignTypes()} as input.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotFrequency(assignedTypesDF, showLegend = FALSE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{assignedTypesDF}] 
A dataframe from the function \code{assignTypes}.

\item[\code{showLegend}] 
Logical indicating whether to show the legend. Default is FALSE.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function visualizes the frequency of assigned genotypes based of the newly generated data.
\end{Details}
%
\begin{Value}
A plot showing the frequency of assigned genotypes.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{assignTypes}{assignTypes}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")

# Run 
queryFastaData <- readFasta(test)
df <- assignTypes(queryFastaData, "p-distance")

plotFrequency(df)
\end{ExampleCode}
\end{Examples}
\HeaderA{plotTree}{Plots a simple phylogenetic tree}{plotTree}
\keyword{genotype}{plotTree}
\keyword{phylogenetics}{plotTree}
%
\begin{Description}
Plots a simple phylogenetic tree using the genetic distances estimated by 
\code{pairwiseDistances()}.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotTree(distance_matrix)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{distance\_matrix}] 
Distance matrix from the function \code{pairwiseDistances}.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function visualizes a phylogenetic tree based on the calculated pairwise genetic distances.
\end{Details}
%
\begin{Value}
A plot object representing the phylogenetic tree.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{pairwiseDistances}{pairwiseDistances}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "input_aln.fasta", package = "rhinotypeR")

# Example usage
fastaD <- readFasta(test)
pdistances <- pairwiseDistances(fastaD, "p-distance")
plotTree(pdistances)
\end{ExampleCode}
\end{Examples}
\HeaderA{readFasta}{Reads sequence alignment/fasta files into R for processing}{readFasta}
\keyword{sequence analysis}{readFasta}
\keyword{data input}{readFasta}
%
\begin{Description}
Reads a sequence alignment or fasta file into R for processing.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
readFasta(fastaFile)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fastaFile}] 
Path to the fasta file. 
A FASTA file is a text-based method for encoding nucleotide or peptide sequences, 
utilizing single-letter codes to represent base pairs or amino acids. 
Each sequence in FASTA format commences with a header line, 
succeeded by lines containing the sequence data.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function reads a FASTA file into R and returns the sequence data for further analysis.
\end{Details}
%
\begin{Value}
A list containing the sequence data from the FASTA file.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{countSNPs}{countSNPs}}, \code{\LinkA{SNPeek}{SNPeek}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "test.fasta", package = "rhinotypeR")

# Run the command
readFasta(test)
\end{ExampleCode}
\end{Examples}
\HeaderA{SNPeek}{Visualize single nucleotide polymorphisms}{SNPeek}
\keyword{SNP}{SNPeek}
\keyword{visualization}{SNPeek}
%
\begin{Description}
Visualizes single nucleotide polymorphisms (SNPs) 
relative to a specified reference sequence. 
To specify the reference, manually move it to the bottom of the alignment. 
Substitutions are color-coded by nucleotide: A = green, T = red, C = blue, G = yellow.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
SNPeek(fastaData, showLegend = FALSE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fastaData}] 
The fasta file used here is the output from the function \code{readFasta}.

\item[\code{showLegend}] 
Logical indicating whether to show the legend. Default is FALSE.

\end{ldescription}
\end{Arguments}
%
\begin{Details}
This function visualizes SNPs in the provided sequence data, 
using a color-coding scheme for different nucleotides.
\end{Details}
%
\begin{Value}
A plot showing the SNPs relative to a user specified reference sequence.
\end{Value}
%
\begin{Author}
Martha Luka, Ruth Nanjala, Wafaa Rashed, Winfred Gatua, Olaitan Awe 
\end{Author}
%
\begin{SeeAlso}
\code{\LinkA{readFasta}{readFasta}}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
# Load the dataset
test <- system.file("extdata", "test.fasta", package = "rhinotypeR")

fastaData <- readFasta(fastaFile = test)
SNPeek(fastaData, showLegend = FALSE)
\end{ExampleCode}
\end{Examples}
\printindex{}
\end{document}
